pact-tests/Pact/Core/Test/LegacySerialiseTests.hs

```diff
@@ -8,21 +8,84 @@
 module Pact.Core.Test.LegacySerialiseTests where
 
 import Test.Tasty (TestTree, testGroup)
-import Test.Tasty.HUnit (assertBool, testCase)
-import Data.Maybe
+import Test.Tasty.HUnit
 import Pact.Core.Serialise.LegacyPact
-import Pact.Core.Guards
-
-tests :: TestTree
-tests = testGroup "Legacy Serialisation"
-  [ testGroup "KeySet"
-    [ testCase "pred: keys-2"   $ assertBool "KeySet decoding failed" (isJust (decodeKeySet "{\"pred\":\"keys-2\",\"keys\":[\"ddd8\",\"ed0\"]}"))
-    , testCase "pred: keys-all" $ assertBool "KeySet decoding failed" (isJust (decodeKeySet "{\"pred\":\"keys-all\",\"keys\":[\"ddd8\",\"ed0\"]}"))
-    , testCase "pred: keys-any" $ assertBool "KeySet decoding failed" (isJust (decodeKeySet "{\"pred\":\"keys-any\",\"keys\":[\"ddd8\",\"ed0\"]}"))
-    , testCase "pred defaults"  $ assertBool "KeySet decoding failed" (maybe False (\k -> KeysAll == _ksPredFun k) (decodeKeySet "{\"pred\":\"keys-all\",\"keys\":[\"ddd8\",\"ed0\"]}"))
-    , testCase "pred invalid" $ assertBool "Accept invalid pred" (isNothing (decodeKeySet "{\"pred\":123,\"keys\":[\"ddd8\",\"ed0\"]}"))
-    , testCase "pred empty" $ assertBool "Accept empty pred" (isNothing (decodeKeySet "{\"pred\":\"\",\"keys\":[\"ddd8\",\"ed0\"]}"))
-    , testCase "pred custom" $ assertBool "Accept custom pred" (isJust (decodeKeySet "{\"pred\":\"ABC\",\"keys\":[\"ddd8\",\"ed0\"]}"))
-    ]
-    -- TODO: Add more test cases.
-  ]
+
+import Control.Monad (forM)
+import System.Directory
+import System.FilePath
+import qualified Data.ByteString as BS
+import qualified Data.Text.IO as T
+import Pact.Core.Test.ReplTests
+import Pact.Core.Persistence
+import Pact.Core.Persistence.MockPersistence
+import Pact.Core.Serialise
+import Pact.Core.Builtin
+import Data.Foldable (traverse_)
+import Control.Lens
+import Pact.Core.Repl.Compile
+import Data.Default
+import Pact.Core.IR.Term
+
+tests :: IO TestTree
+tests = testGroup "Legacy Repl Tests" <$> legacyTests
+
+legacyTestDir :: String
+legacyTestDir = "pact-tests" </> "legacy-serial-tests"
+
+
+-- Load a set of test cases from the file system.
+-- Returns a list of tuples of:
+--    - Test case name
+--    - Test repl scripts with `expect` calls (*.repl)
+--    - Legacy-encoded version of the same module (*.json)
+replTestFiles :: IO [(String, [FilePath], [FilePath])]
+replTestFiles = do
+  base <- getDirectoryContents legacyTestDir
+  forM base $ \p -> do
+    let testPath = legacyTestDir </> p
+    files <- getDirectoryContents testPath
+    let replFiles = filter (isExtensionOf "repl") files
+        modFiles =  filter (isExtensionOf "json") files
+    pure (p, replFiles, modFiles)
+
+legacyTests :: IO [TestTree]
+legacyTests = do
+  t <- replTestFiles
+  forM t $ \(p, repl, lm) -> do
+    -- Perform legacy decoding on all `.json` files for that test
+    lm' <- sequence <$> traverse (toModuleData p) lm
+
+    case lm' of
+      Nothing -> error "Reading existing modules failed"
```

Comment:
Honestly this error is probably fine since we're failing in test prep. However, it does imply that there is a potential failure in the parsing and conversion to module data. If it's not too much  work, we should raise that exception specifically